README
NAME
    Class::InsideOut - a safe, simple inside-out object construction kit

VERSION
    This documentation refers to version 1.10

SYNOPSIS
      package My::Class;
     
  use Class::InsideOut qw( public readonly private register id );
     
  public     name => my %name;    # accessor: name()
      readonly   ssn  => my %ssn;     # read-only accessor: ssn()
      private    age  => my %age;     # no accessor
     
  sub new { register( shift ) }
     
  sub greeting {
        my $self = shift;
        return "Hello, my name is $name{ id $self }";
      }

DESCRIPTION
    This is a simple, safe and streamlined toolkit for building inside-out
    objects. Unlike most other inside-out object building modules already on
    CPAN, this module aims for minimalism and robustness:

    *   Does not require derived classes to subclass it

    *   Uses no source filters, attributes or "CHECK" blocks

    *   Supports any underlying object type including black-box inheritance

    *   Does not leak memory on object destruction

    *   Overloading-safe

    *   Thread-safe for Perl 5.8.5 or better

    *   "mod_perl" compatible

    *   Makes no assumption about inheritance or initializer needs

    It provides the minimal support necessary for creating safe inside-out
    objects and generating flexible accessors.

  Additional documentation
    *   Class::InsideOut::Manual::About -- Guide to the inside-out
        technique, the "Class::InsideOut" philosophy, and other inside-out
        implementations

    *   Class::InsideOut::Manual::Advanced -- Advanced topics including
        customizing accessors, black-box inheritance, serialization and
        thread safety

USAGE
  Importing "Class::InsideOut"
    "Class::InsideOut" automatically imports several critical methods into
    the calling package, including "DESTROY" and support methods for
    serializing objects with "Storable". These methods are intimately tied
    to correct functioning of inside-out objects and will always be imported
    regardless of whether additional functions are requested.

    Additional functions may be imported as usual by including them as
    arguments to "use". For example:

      use Class::InsideOut qw( register public );
     
  public name => my %name;
     
  sub new { register( shift ) }

    As a shortcut, "Class::InsideOut" supports two tags for importing sets
    of functions:

    *   ":std" provides "id", "private", "public", "readonly" and "register"

    *   ":all" imports all functions (including an optional constructor)

    Note: Automatic imports can be bypassed via "require" or by passing an
    empty list to "use Class::InsideOut". There is almost no circumstance in
    which this is a good idea.

  Object properties and accessors
    Object properties are declared with the "public", "readonly" and
    "private" functions. They must be passed a label and the lexical hash
    that will be used to store object properties:

      public   name => my %name;
      readonly ssn  => my %ssn;
      private  age  => my %age;

    Properties for an object are accessed through an index into the lexical
    hash based on the memory address of the object. This memory address
    *must* be obtained via "Scalar::Util::refaddr". The alias "id" may be
    imported for brevity.

      $name{ refaddr $self } = "James";
      $ssn { id      $self } = 123456789;
      $age { id      $self } = 32;

    Tip: since "refaddr" and "id" are function calls, it may be efficient to
    store the value once at the beginning of a method, particularly if it is
    being called repeatedly, e.g. within a loop.

    Object properties declared with "public" will have an accessor created
    with the same name as the label. If the accessor is passed an argument,
    the property will be set to the argument. The accessor always returns
    the value of the property.

      # Outside the class
      $person = My::Class->new;
      $person->name( "Larry" );

    Object properties declared with "readonly" will have a read-only
    accessor created. The accessor will die if passed an argument to set the
    property value. The property may be set directly in the hash from within
    the class package as usual.

      # Inside the class
      $ssn { id $person } = 987654321;
     
  # Inside or outside the class
      $person->ssn( 123456789 );      # dies

    Property accessors may also be hand-written by declaring the property
    "private" and writing whatever style of accessor is desired. For
    example:

      sub age     { $age{ id $_[0] } }
      sub set_age { $age{ id $_[0] } = $_[1] }

    Hand-written accessors will be very slightly faster as generated
    accessors hold a reference to the property hash rather than accessing
    the property hash directly.

    It is also possible to use a package hash instead of a lexical hash to
    store object properties:

      public name => our %name;

    However, this makes private object data accessable outside the class and
    incurs a slight performance penalty when accessing the property hash
    directly; it is not recommended to do this unless you really need it for
    some specialized reason.

  Object construction
    "Class::InsideOut" provides no default constructor method as there are
    many possible ways of constructing an inside-out object. This avoids
    constraining users to any particular object initialization or superclass
    initialization methodology.

    By using the memory address of the object as the index for properties,
    *any* type of reference may be used as the basis for an inside-out
    object with "Class::InsideOut".

      sub new {
        my $class = shift;
     
    my $self = \( my $scalar );    # anonymous scalar
      # my $self = {};                 # anonymous hash
      # my $self = [];                 # anonymous array
      # open my $self, "<", $filename; # filehandle reference
     
    bless $self, $class;
        register( $self );
      }

    However, to ensure that the inside-out object is thread-safe, the
    "register" function *must* be called on the newly created object. The
    "register" function may also be called with just the class name for the
    common case of blessing an anonymous scalar.

      register( $class ); # same as register( bless \(my $s), $class )

    As a convenience, "Class::InsideOut" provides an optional "new"
    constructor for simple objects. This constructor automatically
    initializes the object from key/value pairs passed to the constructor
    for all keys matching the name of a property (including otherwise
    "private" or "readonly" properties).

    A more advanced technique for object construction uses another object,
    usually a superclass object, as the object reference. See "black-box
    inheritance" in Class::InsideOut::Manual::Advanced.

  Object destruction
    "Class::InsideOut" automatically exports a special "DESTROY" function.
    This function cleans up object property memory for all declared
    properties the class and for all "Class::InsideOut" based classes in the
    @ISA array to avoid memory leaks or data collision.

    Additionally, if a user-supplied "DEMOLISH" function is available in the
    same package, it will be called with the object being destroyed as its
    argument. "DEMOLISH" can be used for custom destruction behavior such as
    updating class properties, closing sockets or closing database
    connections. Object properties will not be deleted until after
    "DEMOLISH" returns.

      # Sample DEMOLISH: Count objects demolished (for whatever reason)
     
  my $objects_destroyed;
     
  sub DEMOLISH {
        $objects_destroyed++;
      }

    "DEMOLISH" will only be called if it exists for an object's actual
    class. "DEMOLISH" will not be inherited and "DEMOLISH" will not be
    called automatically for any superclasses.

    "DEMOLISH" should manage any necessary calls to superclass "DEMOLISH"
    methods. As with "new", implementation details are left to the user
    based on the user's approach to object inheritance. Depending on how the
    inheritance chain is constructed and how "DEMOLISH" is being used, users
    may wish to entirely override superclass "DEMOLISH" methods, rely upon
    "SUPER::DEMOLISH", or may prefer to walk the entire @ISA tree:

      use Class::ISA;
     
  sub DEMOLISH {
        my $self = shift;
        # class specific demolish actions
     
    # DEMOLISH for all parent classes, but only once
        my @parents = Class::ISA::super_path( __PACKAGE__ );
        my %called;
        for my $p ( @parents  ) {
          my $demolish = $p->can('DEMOLISH');
          $demolish->($self) if not $called{ $demolish }++;
        }
      }

FUNCTIONS
  "id"
      $name{ id $object } = "Larry";

    This is a shorter, mnemonic alias for "Scalar::Util::refaddr". It
    returns the memory address of an object (just like "refaddr") as the
    index to access the properties of an inside-out object.

  "new"
      My::Class->new( name => "Larry", age => 42 );

    This simplistic constructor is provided as a convenience and is only
    exported on request. When called as a class method, it returns a blessed
    anonymous scalar. Arguments will be used to initialize all matching
    inside-out class properties in the @ISA tree. The argument may be a hash
    or hash reference.

    Note: Properties are set directly, not via accessors. This means
    "set_hook" functions will not be called. For more robust argument
    checking, you will need to implement your own constructor.

  "options"
      Class::InsideOut::options( \%new_options );
      %current_options = Class::InsideOut::options();

    The "options" function sets default options for use with all subsquent
    property definitions for the calling package. If called without
    arguments, this function will return the options currently in effect.
    When called with a hash reference of options, these will be joined with
    the existing defaults, overriding any options of the same name.

  "private"
      private weight => my %weight;
      private haircolor => my %hair_color, { %options };

    This is an alias to "property" that also sets the privacy option to
    'private'. It will override default options or options passed as an
    argument.

  "property"
      property name => my %name;
      property rank => my %rank, { %options };

    Declares an inside-out property. Two arguments are required and a third
    is optional. The first is a label for the property; this label will be
    used for introspection and generating accessors and thus must be a valid
    perl identifier. The second argument must be the lexical hash that will
    be used to store data for that property. Note that the "my" keyword can
    be included as part of the argument rather than as a separate statement.
    The property will be tracked for memory cleanup during object
    destruction and for proper thread-safety.

    If a third, optional argument is provided, it must be a reference to a
    hash of options that will be applied to the property and will override
    any default options that have been set.

  "public"
      public height => my %height;
      public age => my %age, { %options };

    This is an alias to "property" that also sets the privacy option to
    'public'. It will override default options or options passed as an
    argument.

  "readonly"
      readonly ssn => my %ssn;
      readonly fingerprint => my %fingerprint, { %options };

    This is an alias to "property" that sets the privacy option to 'public'
    and adds a "set_hook" option that dies if an attempt is made to use the
    accessor to change the property. It will override default options or
    options passed as an argument.

  "register"
      register( bless( $object, $class ) ); # register the object 
      register( $reference, $class );       # automatic bless 
      register( $class );                   # automatic blessed scalar

    Registers objects for thread-safety. This should be called as part of a
    constructor on a object blessed into the current package. Returns the
    resulting object. When called with only a class name, "register" will
    bless an anonymous scalar reference into the given class. When called
    with both a reference and a class name, "register" will bless the
    reference into the class.

OPTIONS
    Options customize how properties are generated. Options may be set as a
    default with the "options" function or passed as a hash reference to
    "public", "private" or "property".

    Valid options include:

  "privacy"
      property rank => my %rank, { privacy => 'public' };
      property serial => my %serial, { privacy => 'private' };

    If the *privacy* option is set to *public*, an accessor will be created
    with the same name as the label. If the accessor is passed an argument,
    the property will be set to the argument. The accessor always returns
    the value of the property.

  "get_hook"
      public list => my %list, {
          get_hook => sub { @$_ }
      };

    Defines an accessor hook for when values are retrieved. $_ is locally
    aliased to the property value for the object. *The return value of the
    hook is passed through as the return value of the accessor.* See
    "Customizing Accessors" in Class::InsideOut::Manual::Advanced for
    details.

  "set_hook"
      public age => my %age, {
         set_hook => sub { /^\d+$/ or die "must be an integer" }
      };

    Defines an accessor hook for when values are set. The hook subroutine
    receives the entire argument list. $_ is locally aliased to the first
    argument for convenience. The property receives the value of $_. See
    "Customizing Accessors" in Class::InsideOut::Manual::Advanced for
    details.

SEE ALSO
    Programmers seeking a more full-featured approach to inside-out objects
    are encouraged to explore Object::InsideOut. Other implementations are
    also noted in Class::InsideOut::Manual::About.

KNOWN LIMITATIONS
    Requires weak reference support (Perl >= 5.6) and Scalar::Util::weaken()
    to avoid memory leaks and to provide thread-safety.

ROADMAP
    Features slated for after the 1.0 release include:

    *   Adding support for Data::Dump::Streamer serialization hooks

    *   Adding additional accessor styles (e.g. get_name()/set_name())

    *   Further documentation revisions and clarification

BUGS
    Please report bugs or feature requests using the CPAN Request Tracker:
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Class-InsideOut>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    David A. Golden (DAGOLDEN)

COPYRIGHT AND LICENSE
    Copyright (c) 2006, 2007 by David A. Golden

    Licensed under the Apache License, Version 2.0 (the "License"); you may
    not use this file except in compliance with the License. You may obtain
    a copy of the License at L<http://www.apache.org/licenses/LICENSE-2.0>

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



Changes
Revision history for Perl module Class::InsideOut

{{$NEXT}}

1.14      2017-04-02 13:52:18-04:00 America/New_York

    Fixed:

    - Passes tests when '.' is not in @INC.

1.13      2013-04-25 13:59:42 America/New_York

    Added:
    
    - allow set_hook and get_hook to be blessed or overloaded objects

1.12      2013-04-24 21:02:39 America/New_York

    Docs:
    - Typos fixed

    Other:
    - Issue tracker is now GitHub

1.11      2013-01-23 21:33:13 America/New_York

    Bug fixes:
    - fixed tests for recent, XS-only Scalar::Util

    Other:
    - converted distribution management to Dist::Zilla

1.10 Mon Aug 24 20:44:46 EDT 2009

    Bug fixes:
    - changed method of calling Exporter::import() to avoid confusing DProf
    - fixed detection of missing 'weaken' function in newer Scalar::Util
      (fixes test t/15_no_weaken_fallback.t) (RT#47623)

    Other: 
    - cleanup: eliminate test warning from bogus superclass in a test class
    - docs: Noted that set_hooks aren't called by new() (RT#48106)

1.09 Fri Jan  4 18:42:41 EST 2008

    - bugfix: allow use of upper or mixed case property accessors 
      (David Schmitt)

    - testfix: tests will now pass on Perl 5.005 -- though 5.005 is not
      recommended due to the lack of weak references
    
    - pod: added a note about using "our" instead of "my" for properties 

    - critic: various cleanups to about Perl::Critic level 5
    
    - moved author tests to xt directory for CPANTS compatibility

1.08 Thu Aug 23 07:12:33 EDT 2007
    
    - STORABLE_attach warns instead of dying if it can't provide a singleton
      back to Storable, leaving Storable to croak instead; on 5.6.2 (and 
      possibly older Perls), this change works around Storable dying during
      cleanup if STORABLE_attach dies inside an eval()
    
    - thread tests skip on perl < 5.8.5 due to unexplained thread failures;
      perl585delta alludes to thread fixes when weak references are in use
      and this may be related
    
    - removed "use warnings" from singleton test files for back compatibility
    
    - moved pod/coverage tests to t_extra/ and stopped depending on an
      environment variable to allow them to run
    
    - changed to the Apache License, version 2.0; (it's clearer, relicensable,
      and is explicit about contributions)
    
1.07 Fri Aug 10 07:41:00 EDT 2007
    
    - fixes for using ":storable" with older versions of Exporter
      that required tags to come first. Now we just strip it during import.

1.06 Mon Feb 12 19:34:59 EST 2007
    
    - all tests involving Storable now skip if Storable isn't installed
      (instead of only a few) (David Cantrell)
    
1.05 Sun Feb 11 16:29:16 EST 2007
    
    - fixed optional new method with hash reference (RT#24839)

1.04 Thu Jan 18 21:47:19 EST 2007
    
    - added 'readonly' accessor-creator

1.0301 Wed Nov 15 06:08:47 EST 2006
    
    - removing email address for bug reports to (hopefully)
      cut down on RT ticket spam

1.03 Wed Oct 11 10:55:28 EDT 2006
    
    - thread tests skip if thread creation fails
    
    - added an examples directory with a reference to some of the objects
      used in testing 

1.02 Tue Aug 15 09:26:11 EDT 2006
    
    - fixed duplicate property name checking (RT#20997)

1.01 Thu Jul 27 12:30:54 EDT 2006
    
    - register() now handles standard cases with "sufficiently advanced
      technology" and does the right thing
 
    - optional simple "new" constructor added
    
    - changed "foreign" inheritance to "black-box" inheritance in Pod

1.00 Fri May 12 21:25:34 EDT 2006
    
    - removed Data::Dump::Streamer support attempts until RT#19060 can
      be addressed
      
0.90_02 Thu May  4 00:56:04 EDT 2006
    
    - fixed up META.yml and a pod typo

0.90_01 Wed May  3 20:47:34 EDT 2006

    - *API CHANGE* renamed serialization hooks to FREEZE and THAW
    
    - added support for STORABLE_attach for singletons
    
    - improved test coverage
    
    - refactored some of the test classes
    
    - refactored STORABLE_* code
    
    - refactored documentation
    
    - added draft (non-working) support for serialization with
      Data::Dump::Streamer

0.14 Thu Mar 16 23:07:22 EST 2006

    - Fixed test bug where thread test wouldn't properly skip for Win32 Perl
      5.6 (with ithreads enabled for pseudofork but no threads.pm)
    
    - Improved handling of "die" in set_hook and get_hook callbacks

0.13 Fri Mar 10 10:32:08 EST 2006

    - Test for Scalar::Util::weaken fallback had a bug that would fail on
      Perl versions prior to 5.6.  Fixed to be fully backwards compatible. 
    
0.12 Wed Mar  8 21:58:58 EST 2006
    
    - Added workaround for unavailable Scalar::Util::weaken; gives warning
      about not running thread-safe; added missing weaken skips to thread/fork
      test files

    - property/public/private arguments are now validated.  Property names must
      be identifiers; Property options must be given as hash references;
      Duplicate property names are not permitted. 

    - options argument checked for proper format; option values are checked
      for validity

    - register argument checked for blessed reference

    - added optional pod/pod_coverage tests (skipped by default)

0.11 Sat Jan 28 11:09:50 EST 2006

    - tweaked Build.PL; dropped Test::More dependency to 0.45 to help pass
      ActiveState automated tests (0.45 first to offer thread safety and in the
      5.008 core); fixed copy/paste error for cleanup

    - documentation cleanup; fixed links; restructured Pod for options;
      softened alpha warnings as we get closer to a stable API

0.10 Fri Jan 27 01:41:38 EST 2006

    - added "set_hook" and "get_hook" option for custom accessor/mutator
      manipulations

    - added ":all" and ":std" tags

    - cleaned up or reorganized documentation in places; revised synopsis
      to be shorter

0.09 Fri Jan 20 14:55:14 EST 2006


t/Util.pm
package t::Util;
use strict;
use warnings;
use Exporter 'import';
our @EXPORT = qw( 
    test_constructor TC
    test_accessors TA
    test_methods TM
    test_new TN
);

use Test::More;

sub load_fail_msg { return "because $_[0] isn't loaded" }
sub method_fail_msg { return "because $_[0] can't $_[1]" }

sub TC { return 1 + TN() }
sub test_constructor {
    my ($class, @args) = @_;
    my $pass = require_ok( $class );
    my $o;
    SKIP: {
        skip load_fail_msg($class), TC() - 1 unless $pass;
        $o = test_new($class,@args);
    }
    return $o;
}

sub TN { return 2 }
sub test_new {
    my ($class, @args) = @_;
    my $o;
    ok( $o = $class->new(@args), "... creating a $class object");
    ok( $o->isa($class), "... confirming object is a $class" );
    return $o;
}


sub TA { return TN() + 7 }
sub test_accessors {
    my ($o, $prop, $prefix) = @_;
    my $class = ref($o);
    $prefix ||= { get => q{}, set => 'set_' };
    my $acc = $prefix->{get} . $prop;
    my $mut = $prefix->{set} . $prop;
    my $pass = ok( $o->can($acc), "... found accessor function '$acc'" );
    $pass = ok( $o->can($mut), "... found mutator function '$mut'" ) && $pass;

    SKIP: {
        skip load_fail_msg($class), TA() - 2  unless $pass;
        my $p = test_new($class);
        my $value1 = "foo";
        my $value2 = "bar";
        if ( $acc ne $mut ) {
            is( $o->$mut($value1), $o, 
                "... $mut(\$value1) returns self for object 1" );
            is( $o->$acc, $value1,
                "... $acc() equals \$value1 for object 1" );
            is( $p->$mut($value2), $p, 
                "... $mut(\$value2) returns self for object 2" );
            is( $p->$acc, $value2,
                "... $acc() equals \$value2 for object 2" );
            is( $o->$acc, $value1,
                "... $acc() still equals \$value1 for object1" );
        }
        else {
            is( $o->$mut($value1), $value1, 
                "... $mut(\$value1) returns \$value1 for object 1" );
            is( $o->$acc, $value1,
                "... $acc() equals \$value1 for object 1" );
            is( $p->$mut($value2), $value2, 
                "... $mut(\$value2) returns \$value2 for object 2" );
            is( $p->$acc, $value2,
                "... $acc() equals \$value2 for object 2" );
            is( $o->$acc, $value1,
                "... $acc() still equals \$value1 for object1" );
        }
    }
    return $pass;
}

sub TM { return 2 }
sub test_methods {
    my ($o, $case) = @_;
    my ($method, $args, $result) = @$case;
    my $class = ref($o);
    my $pass = can_ok( $o, $method );
    SKIP: {
        skip method_fail_msg($class, $method), TM() - 1  unless $pass;
        is( $o->$method(@$args), $result, "$method gave correct result" );
    }
    return $pass;
}

1;


t/Object/Hash.pm
package t::Object::Hash;
use strict;

use Class::InsideOut qw( public register );

public name => my %name; 
public weight => my %weight;

sub new { register( bless \(my $s), shift ) }

1;


t/Object/Array.pm
package t::Object::Array;
use strict;

use Class::InsideOut qw( public register );

public name => my %name; 
public height => my %height;

sub new { register( bless \(my $s), shift ) }

1;


Build.PL
use strict;
use Module::Build;

my $class = Module::Build->subclass(
    class => "Module::Build::WikiDoc",
    code => <<'SUBCLASS', );

    sub ACTION_wikidoc {
        my $self = shift;
        eval "use Pod::WikiDoc";
        if ( $@ eq '' ) {
            my $parser = Pod::WikiDoc->new( {
                comment_blocks => 1, 
                keywords => { VERSION => $self->dist_version },
            });
            for my $src ( keys %{ $self->find_pm_files() } ) {
                (my $tgt = $src) =~ s{\.pm$}{.pod};
                $parser->filter( {
                    input   => $src,
                    output  => $tgt,
                });
                print "Creating $tgt\n";
                $tgt =~ s{\\}{/}g;
                $self->_add_to_manifest( 'MANIFEST', $tgt );
            }
        }
        else {
            warn "Pod::WikiDoc not available. Skipping wikidoc.\n";
        }
    }

    sub ACTION_test {
        my $self = shift;
        my $missing_pod;
        for my $src ( keys %{ $self->find_pm_files() } ) {
            (my $tgt = $src) =~ s{\.pm$}{.pod};
            $missing_pod = 1 if ! -e $tgt;
        }
        if ( $missing_pod ) {
            $self->depends_on('wikidoc');
            $self->depends_on('build');
        }
        $self->SUPER::ACTION_test;
    }

    sub ACTION_testpod {
        my $self = shift;
        $self->depends_on('wikidoc');
        $self->SUPER::ACTION_testpod;
    }

    sub ACTION_distmeta {
        my $self = shift;
        $self->depends_on('wikidoc');
        $self->SUPER::ACTION_distmeta;
    }

SUBCLASS

my $builder = $class->new( 
    module_name         => 'Class::InsideOut',
    dist_author         => 'David A. Golden <dagolden@cpan.org>',
    license             => 'apache',
    create_readme       => 1,
    create_makefile_pl  => 'traditional',
    requires        => {
        'perl'           => 5.005, # says Perl::MinimumVersion
        'base'           => 0,
        'Class::ISA'     => 0, # why reimplement what works
        'Config'         => 0,
        'Exporter'       => 0,
        'Test::More'     => 0.45, # thread-safe
        'Scalar::Util'   => 1.09, # refaddr
    },
    add_to_cleanup      => [ 'Class-InsideOut-*' ],
    meta_add            => { 
        no_index => { 
            file => [ qw{ 
                lib/Class/InsideOut/Manual/About.pm
                lib/Class/InsideOut/Manual/Advanced.pm 
            } ],
            directory => [ 'examples' ],
        } 
    },

);

$builder->create_build_script;


lib/Class/InsideOut/Manual/About.pm
# PODNAME: Class::InsideOut::Manual::About

=begin wikidoc

= NAME

Class::InsideOut::Manual::About - guide to this and other implementations of the
inside-out technique

= VERSION

This documentation refers to version %%VERSION%%

= DESCRIPTION

This manual provides an overview of the inside-out technique and its
application within {Class::InsideOut} and other modules.  It also provides a
list of references for further study.

== Inside-out object basics

Inside-out objects use the blessed reference as an index into lexical data
structures holding object properties, rather than using the blessed reference
itself as a data structure.

 $self->{ name }        = "Larry"; # classic, hash-based object
 $name{ refaddr $self } = "Larry"; # inside-out

The inside-out approach offers three major benefits:

* Enforced encapsulation: object properties cannot be accessed directly
from ouside the lexical scope that declared them
* Making the property name part of a lexical variable rather than a hash-key
means that typos in the name will be caught as compile-time errors (if
using [strict])
* If the memory address of the blessed reference is used as the index,
the reference can be of any type

In exchange for these benefits, robust implementation of inside-out
objects can be quite complex.  {Class::InsideOut} manages that complexity.

== Philosophy of {Class::InsideOut}

{Class::InsideOut} provides a set of tools for building safe inside-out classes
with maximum flexibility.

It aims to offer minimal restrictions beyond those necessary for robustness of
the inside-out technique.  All capabilities necessary for robustness should be
automatic.  Anything that can be optional should be.  The design should not
introduce new restrictions unrelated to inside-out objects, such as attributes
and {CHECK} blocks that cause problems for {mod_perl} or the use of source
filters for syntatic sugar.

As a result, only a few things are mandatory:

* Properties must be based on hashes and declared via {property}
* Property hashes must be keyed on the {Scalar::Util::refaddr}
* {register} must be called on all new objects

All other implementation details, including constructors, initializers and
class inheritance management are left to the user (though a very simple
constructor is available as a convenience).  This does requires some additional
work, but maximizes freedom.  {Class::InsideOut} is intended to be a base class
providing only fundamental features.  Subclasses of {Class::InsideOut} could be
written that build upon it to provide particular styles of constructor,
destructor and inheritance support.

== Other modules on CPAN

* [Object::InsideOut] -- This is perhaps the most full-featured, robust
implementation of inside-out objects currently on CPAN.  It is highly
recommended if a more full-featured inside-out object builder is needed.
Its array-based mode is faster than hash-based implementations, but black-box
inheritance is handled via delegation, which imposes certain limitations.

* [Class::Std] -- Despite the name, this does not reflect currently known best
practices for inside-out objects.  Does not provide thread-safety with CLONE
and doesn't support black-box inheritance.  Has a robust
inheritance/initialization system.

* [Class::BuildMethods] -- Generates accessors with encapsulated storage using
a flyweight inside-out variant. Lexicals properties are hidden; accessors must
be used everywhere. Not thread-safe.

* [Lexical::Attributes] -- The original inside-out implementation, but missing
some key features like thread-safety.  Also, uses source filters to provide
Perl-6-like object syntax. Not thread-safe.

* [Class::MakeMethods::Templates::InsideOut] -- Not a very robust
implementation. Not thread-safe.  Not overloading-safe.  Has a steep learning
curve for the Class::MakeMethods system.

* [Object::LocalVars] -- My own original thought experiment with 'outside-in'
objects and local variable aliasing. Not safe for any production use and offers
very weak encapsulation.

== References for further study

Much of the Perl community discussion of inside-out objects has taken place on
Perlmonks ([http://perlmonks.org]).  My scratchpad there has a fairly
comprehensive list of articles
([http://perlmonks.org/index.pl?node_id=360998]).  Some of the more
informative articles include:

* Abigail-II. "Re: Where/When is OO useful?". July 1, 2002.
[http://perlmonks.org/index.pl?node_id=178518]
* Abigail-II. "Re: Tutorial: Introduction to Object-Oriented Programming".
December 11, 2002. [http://perlmonks.org/index.pl?node_id=219131]
* demerphq. "Yet Another Perl Object Model (Inside Out Objects)". December 14,
2002. [http://perlmonks.org/index.pl?node_id=219924]
* xdg. "Threads and fork and CLONE, oh my!". August 11, 2005.
[http://perlmonks.org/index.pl?node_id=483162]
* jdhedden. "Anti-inside-out-object-ism". December 9, 2005.
[http://perlmonks.org/index.pl?node_id=515650]

= SEE ALSO

* [Class::InsideOut]
* [Class::InsideOut::Manual::Advanced]

= AUTHOR

David A. Golden (DAGOLDEN)

= COPYRIGHT AND LICENSE

Copyright (c) 2006, 2007 by David A. Golden

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 
L<http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

=end wikidoc



Makefile.PL
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.17;

my %WriteMakefileArgs = (
  "ABSTRACT" => "a safe, simple inside-out object construction kit",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.17"
  },
  "DISTNAME" => "Class-InsideOut",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "Class::InsideOut",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Class::ISA" => 0,
    "Exporter" => 0,
    "Scalar::Util" => "1.09",
    "Storable" => 0,
    "overload" => 0,
    "strict" => 0,
    "vars" => 0
  },
  "TEST_REQUIRES" => {
    "DynaLoader" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::File" => 0,
    "Test::More" => "0.45",
    "XSLoader" => 0,
    "lib" => 0,
    "threads" => 0,
    "warnings" => 0
  },
  "VERSION" => "1.15",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Class::ISA" => 0,
  "DynaLoader" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::File" => 0,
  "Scalar::Util" => "1.09",
  "Storable" => 0,
  "Test::More" => "0.45",
  "XSLoader" => 0,
  "lib" => 0,
  "overload" => 0,
  "strict" => 0,
  "threads" => 0,
  "vars" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
